### PR TITLE
#83: Find installations by script names

### PR DESCRIPTION
**Pull request by kaklakariada**

### CODE REVIEW

**Comment by ckunki** [comment]
Failed to execute goal org.sonatype.ossindex.maven:ossindex-maven-plugin:3.2.0:audit (audit) on project s3-document-files-virtual-schema: Detected 1 vulnerable components:
* [org.eclipse.jetty:jetty-client:jar:9.4.43.v20210629](https://ossindex.sonatype.org/component/pkg:maven/org.eclipse.jetty/jetty-client@9.4.43.v20210629?utm_source=ossindex-client&utm_medium=integration&utm_content=1.8.1) in compile
    * CVE-2022-2047, severity CWE-20: Improper Input Validation (2.7)

File: extension/src/extension.ts [hunk 1]
```diff
@@ -1,32 +1,35 @@
...
+import { findInstallations } from "./implementation";
 
 export function createExtension(): ExasolExtension {
     const version = CONFIG.version;
-    const filename = CONFIG.fileName;
+    const fileName = CONFIG.fileName;
     const fileSize = CONFIG.fileSizeBytes;
     const repoBaseUrl = "https://github.com/exasol/s3-document-files-virtual-schema"
-    const downloadUrl = `${repoBaseUrl}/releases/download/${version}/${filename}`;
+    const downloadUrl = `${repoBaseUrl}/releases/download/${version}/${fileName}`;
```
Comment (ckunki) [inline extension/src/extension.ts:18]: Am I correct, that repoBaseUrl is used only for licenseUrl? 
If so then IMHO we could as well just create a variable ``licenseUrl`` and just use it.

File: extension/src/extension.ts [hunk 2]
```diff
@@ -1,32 +1,35 @@
...
     registerExtension,
     SqlClient
 } from "@exasol/extension-manager-interface";
+import { Context } from "./common";
 import { CONFIG } from "./extension-config";
+import { findInstallations } from "./implementation";
 
 export function createExtension(): ExasolExtension {
     const version = CONFIG.version;
-    const filename = CONFIG.fileName;
```
Comment (ckunki) [inline extension/src/extension.ts:13]: Alternatively we could maybe 
* drop these individual variables 
* put CONFIG.version and .filename immediately into context
* always access filename and version through context: ``context.version``, ``context.filename``

File: extension/src/implementation.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    const importScript = findScriptByName(scripts, IMPORT_SCRIPT_NAME);
+    const adapterScript = findScriptByName(scripts, ADAPTER_SCRIPT_NAME);
+    if (isValidImportScript(importScript) && isValidAdapterScript(adapterScript)) {
+        return [{
+            name: `${adapterScript.schema}.${adapterScript.name}`, version: context.version, instanceParameters: []
+        }];
+    } else {
+        return [];
+    }
+}
```
Comment (ckunki) [inline extension/src/implementation.ts:19]: I do not understand the role of parameter `context` here.
Maybe you could explain to me?

File: extension/src/implementation.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+function isValidAdapterScript(script: ExaAllScriptsRow): script is ExaAllScriptsRow {
+    if (!script) {
+        console.log(`Adapter script ${ADAPTER_SCRIPT_NAME} not found`)
```
Comment (ckunki) [inline extension/src/implementation.ts:23]: Should this really be written to console, here?
Or is console visible to the human user?

File: src/test/java/com/exasol/adapter/document/files/extension/ExtensionIT.java [hunk 5]
```diff
@@ -49,4 +67,37 @@ void listInstallationsEmpty() {
                 .getInstallations();
         assertThat(installations, hasSize(0));
     }
+
+    @Test
+    void listInstallations_ignoresWrongScriptNames() {
+        final String currentProjectVersion = MavenProjectVersionGetter.getCurrentProjectVersion();
+        createAdapter("wrong_adapater_name", "wrong_import_script_name");
```
Comment (ckunki) [inline src/test/java/com/exasol/adapter/document/files/extension/ExtensionIT.java:74]:
```suggestion
        createAdapter("wrong_adapter_name", "wrong_import_script_name");
```

File: src/test/java/com/exasol/adapter/document/files/extension/ExtensionIT.java [hunk 6]
```diff
@@ -49,4 +67,37 @@ void listInstallationsEmpty() {
...
+    @Test
+    void listInstallations_findsMatchingScripts() {
+        final String currentProjectVersion = MavenProjectVersionGetter.getCurrentProjectVersion();
+        createAdapter("S3_FILES_ADAPTER", "IMPORT_FROM_S3_DOCUMENT_FILES");
+        final List<RestAPIInstallationsResponseInstallation> installations = extensionManagerSetup.client()
+                .getInstallations();
+        assertAll(() -> assertThat(installations, hasSize(1)), //
+                () -> assertThat(installations.get(0).getName(),
+                        equalTo(ExtensionManagerSetup.EXTENSION_SCHEMA_NAME + ".S3_FILES_ADAPTER")),
+                () -> assertThat(installations.get(0).getVersion(), equalTo(currentProjectVersion)),
```
Comment (ckunki) [inline src/test/java/com/exasol/adapter/document/files/extension/ExtensionIT.java:89]: I did not fully understand the role of version here.
Maybe you could explain?

File: src/test/java/com/exasol/adapter/document/files/extension/ExtensionManagerInstaller.java [hunk 7]
```diff
@@ -7,51 +7,82 @@
```
Comment (ckunki) [inline src/test/java/com/exasol/adapter/document/files/extension/ExtensionManagerInstaller.java:7]: Maybe we should spend a dedicated package "installer" here and create 3 files? 
* interface ExtensionManagerInstaller 
* GoInstaller 
* LocalProjectDirInstaller

File: src/test/java/com/exasol/adapter/document/files/extension/ExtensionTestConfig.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            final Properties props = new Properties();
+            props.load(stream);
+            return props;
+        } catch (final IOException e) {
+            throw new UncheckedIOException("Error reading config file " + configFile, e);
+        }
+    }
+
+    public Optional<Path> getLocalExtensionManagerProject() {
+        return getOptionalValue("localExtensionManager") //
```
Comment (ckunki) [inline src/test/java/com/exasol/adapter/document/files/extension/ExtensionTestConfig.java:39]: I am not sure if functional programming style offers significant benefits in this case 😃 ...

**Review by ckunki — CHANGES_REQUESTED** [review]
Additionally there is a vulnerability finding reported by ossindex 
* [org.eclipse.jetty:jetty-client:jar:9.4.43.v20210629](https://ossindex.sonatype.org/component/pkg:maven/org.eclipse.jetty/jetty-client@9.4.43.v20210629?utm_source=ossindex-client&utm_medium=integration&utm_content=1.8.1) in compile
    * CVE-2022-2047, severity CWE-20: Improper Input Validation (2.7)

File: extension/src/implementation.ts (hunk 4 above)
Comment (kaklakariada) [reply extension/src/implementation.ts:23]: This appears only in the stdout of the server process, not the browser :)

File: src/test/java/com/exasol/adapter/document/files/extension/ExtensionManagerInstaller.java (hunk 7 above)
Comment (kaklakariada) [reply src/test/java/com/exasol/adapter/document/files/extension/ExtensionManagerInstaller.java:7]: ✅

File: src/test/java/com/exasol/adapter/document/files/extension/ExtensionIT.java (hunk 5 above)
Comment (kaklakariada) [reply src/test/java/com/exasol/adapter/document/files/extension/ExtensionIT.java:74]: ✅

**Review by ckunki — APPROVED** [review]

**Review by ckunki — APPROVED** [review]

**Pull request merged**